### PR TITLE
[WIP] add round tripping tests to each group

### PR DESCRIPTION
**Pull request by deads2k**
To properly separate schemes you end up needing to test each API group by itself *and* test your overall scheme.  We've built helpers to make this easy, but our default registration requires consumers to know exactly how to wire all dependent groups which doesn't make sense.

This updates just `apps` to be tested individually and demonstrates the pitfalls.

@kubernetes/sig-apimachinery-pr-reviews @lavalamp @sttts @smarterclayton

### CODE REVIEW

File: pkg/api/fuzzer.go [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+					f.SetInt(1)
+				}
+			}
+		},
+		func(ev *EnvVar, c fuzz.Continue) {
+			ev.Name = c.RandString()
+			if c.RandBool() {
+				ev.Value = c.RandString()
+			} else {
+				c.FuzzNoCustom(ev) // fuzz self without calling this function again
```
Comment (deads2k) [inline pkg/api/fuzzer.go]: The only change in this block.  It used to special case itself and now it doesn't.

File: pkg/apis/apps/register.go [hunk 2]
```diff
@@ -17,13 +17,15 @@ limitations under the License.
...
+	"k8s.io/apimachinery/pkg/apimachinery/announced"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/api"
 )
 
 var (
 	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
-	AddToScheme   = SchemeBuilder.AddToScheme
+	AddToScheme   = announced.CombineSchemeBuilders(SchemeBuilder.AddToScheme, api.AddToScheme).AddToScheme
```
Comment (deads2k) [inline pkg/apis/apps/register.go]: This was neat.  "apps" includes types from the core API, but did not include those types for itself and did not include conversions.  This meant that if you built a scheme with just apps, you'd be unable to roundtrip your types properly.

**Comment by deads2k** [comment]
Wow.  Well clearly we're not ready for this to work.

File: pkg/api/fuzzer.go (hunk 1 above)
Comment (sttts) [inline pkg/api/fuzzer.go]: Looks good enough.

Comment (sttts) [inline pkg/api/fuzzer.go]: Alternatively, put in the old code restricted to the core apigroup.

File: staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go [hunk 3]
```diff
@@ -78,7 +79,17 @@ func RoundTripProtobufTestForScheme(t *testing.T, scheme *runtime.Scheme, fuzzin
 	RoundTripTypes(t, scheme, codecFactory, fuzzer, nil)
 }
 
-var FuzzIters = flag.Int("fuzz-iters", 20, "How many fuzzing iterations to do.")
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip.go]: this is used 8 times through out the code base. Either change all or none.

File: pkg/apis/apps/v1beta1/register.go [hunk 4]
```diff
@@ -35,7 +37,7 @@ func Resource(resource string) schema.GroupResource {
 
 var (
 	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes, addDefaultingFuncs, addConversionFuncs)
-	AddToScheme   = SchemeBuilder.AddToScheme
+	AddToScheme   = announced.CombineSchemeBuilders(SchemeBuilder.AddToScheme, v1.AddToScheme).AddToScheme
```
Comment (sttts) [inline pkg/apis/apps/v1beta1/register.go]: please comment why. Which types are needed?

File: pkg/apis/apps/register.go (hunk 2 above)
Comment (sttts) [inline pkg/apis/apps/register.go]: Please comment why. Which types are needed?

File: pkg/apis/apps/v1beta1/register.go (hunk 4 above)
Comment (deads2k) [inline pkg/apis/apps/v1beta1/register.go]: > please comment why. Which types are needed?

It includes at least a pod spec, which ripples out.  It's neat.  We die on duplicate registration of identical types, so this is hosed for 1.6.

Comment (sttts) [inline pkg/apis/apps/v1beta1/register.go]: With https://github.com/kubernetes/kubernetes/pull/41788? Why should we die on that? Only on different types for same gvk.

Comment (deads2k) [inline pkg/apis/apps/v1beta1/register.go]: > With #41788? Why should we die on that? Only on different types for same gvk.

I just wasn't planning on chasing it and fixing it with my few days before freeze.

**Pull request closed without merging**